src/dbnode/storage/series/buffer.go

```diff
@@ -473,119 +440,346 @@ func (b *dbBuffer) FetchBlocksMetadata(
 ) block.FetchBlockMetadataResults {
 	blockSize := b.opts.RetentionOptions().BlockSize()
 	res := b.opts.FetchBlockMetadataResultsPool().Get()
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if !bucket.canRead() {
-			return
+
+	for _, blockStart := range b.inOrderBlockStarts {
+		if !blockStart.Before(end) || !start.Before(blockStart.Add(blockSize)) {
+			continue
 		}
-		if !start.Before(bucket.start.Add(blockSize)) || !bucket.start.Before(end) {
-			return
+
+		bv, exists := b.bucketVersionsAt(blockStart)
+		if !exists {
+			// Invariant violated. This means the keys in the bucket map does
+			// not match the sorted keys cache, which should never happen.
+			log := b.opts.InstrumentOptions().Logger()
+			log.Errorf(errBucketMapCacheNotInSync)
+			return nil
```

Comment:
Should return an error rather than returning nil here.

Reply:
I guess I've always questioned why we had these "should never happen" invariant violations logs as opposed to returning an error and dealing with it in the application. There are [other places](https://github.com/m3db/m3/blob/master/src/dbnode/storage/shard.go#L385) in which we log and don't return an error. Thoughts? Do you think it's better to change the function return and deal with the invariant violation here?

Reply:
Those places there is no way to actually return the error because the structure of the code (i.e. OnRetrieveBlock is an internal call for caching a block, so there is no RPC to ultimately return an error to at the top level of the calling method).

The invariant pattern is just so that if screwed up stuff actually does happen we can alert off of it.  We should still always surface an error to the caller as best as possible, rather than silently letting the operation finish.

Reply:
Cool, changed.